amethyst_core/src/entity_uuid.rs

```diff
@@ -0,0 +1,157 @@
+    fn new() -> Self {
+        Self {
+            uuid: Uuid::new_v4(),
+        }
+    }
+
+    /// Create a new instance with a pre-defined Uuid, typically the Uuid would be deserialized from the disk or network.
+    fn new_from_uuid(uuid: Uuid) -> Self {
+        Self { uuid }
+    }
+
+    /// Retrieve the Uuid structure contained in this component.
+    pub fn uuid(&self) -> &Uuid {
+        &self.uuid
+    }
+}
+
+/// An ECS resource which allows you to create new EntityUuid components, and later retrieve the entities with the contained UUID.
+#[derive(Debug, Default)]
+pub struct EntityUuidMap {
+    map: FnvHashMap<Uuid, Entity>,
+}
+
+impl EntityUuidMap {
+    /// Create an empty instance of this resource.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Create a new EntityUuid component with a new randomly generated Uuid.
+    pub fn new_uuid(&mut self, e: Entity) -> EntityUuid {
+        let r = EntityUuid::new();
+        self.map.insert(r.uuid, e);
+        r
+    }
+
+    /// Create a new EntityUuid component with a pre-defined Uuid, typically the Uuid would be deserialized from the disk or network.
+    pub fn with_uuid(&mut self, uuid: Uuid, e: Entity) -> EntityUuid {
+        self.map.insert(uuid, e);
+        EntityUuid::new_from_uuid(uuid)
+    }
+
+    /// Retrieve the entity associated with this Uuid, if any.
+    pub fn fetch_entity(&self, uuid: &Uuid) -> Option<Entity> {
+        self.map.get(uuid).cloned()
+    }
+}
+
+/// This system removes unused Uuid<->Entity mappings from the map.
+#[derive(Debug)]
+pub struct EntityUuidSystem {
+    reader: Option<ReaderId<ComponentEvent>>,
+    removed: BitSet,
+}
+
+impl EntityUuidSystem {
+    /// Create a new instance of this system.
+    pub fn new() -> Self {
+        Self {
+            reader: None,
+            removed: BitSet::new(),
+        }
+    }
+}
+
+impl<'s> System<'s> for EntityUuidSystem {
+    type SystemData = (
+        ReadStorage<'s, EntityUuid>,
+        Write<'s, EntityUuidMap>,
+        Entities<'s>,
+    );
+
+    fn setup(&mut self, res: &mut Resources) {
+        Self::SystemData::setup(res);
+        self.reader = Some(
+            res.fetch_mut::<WriteStorage<'_, EntityUuid>>()
+                .channel_mut()
+                .register_reader(),
+        );
+    }
+
+    fn run(&mut self, (storage, mut map, entities): Self::SystemData) {
+        self.removed.clear();
+        let self_events_id = &mut self
+            .reader
+            .as_mut()
+            .expect("EntityUuidSystem::setup was not called before EntityUuidSystem::run");
+        for event in storage.channel().read(self_events_id) {
+            match event {
+                ComponentEvent::Removed(id) => {
+                    self.removed.add(*id);
+                }
+                _ => {}
+            }
+        }
+        for entity in (&self.removed).iter().map(|i| entities.entity(i)) {
+            map.map.retain(|_, &mut e| e != entity);
```

Comment:
Is it possible to use HashMap::remove here?

Reply:
No, unfortunately. Once the Uuid component is removed I have no way to get at the Uuid. So I am only able to remove by the value of the hashmap, rather than the key.